### PR TITLE
Eliminate duplicate allocation of nested string columns

### PR DESCRIPTION
**Pull request by vuule**
## Description
Issue https://github.com/rapidsai/cudf/issues/14965

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/parquet/reader_impl.cpp [hunk 1]
```diff
@@ -156,7 +156,8 @@ void reader::impl::decode_page_data(size_t skip_rows, size_t num_rows)
         valids[idx] = out_buf.null_mask();
         data[idx]   = out_buf.data();
         // only do string buffer for leaf
-        if (out_buf.string_size() == 0 && col_sizes[pass.chunks[c].src_col_index] > 0) {
+        if (idx == max_depth - 1 and out_buf.string_size() == 0 and
+            col_sizes[pass.chunks[c].src_col_index] > 0) {
```
Comment (nvdbaranec) [inline cpp/src/io/parquet/reader_impl.cpp:160]: Recommend renaming this to `col_string_sizes` for clarity.

**Review by ttnghia — APPROVED** [review]

File: cpp/src/io/parquet/reader_impl.cpp [hunk 2]
```diff
@@ -281,7 +282,7 @@ void reader::impl::decode_page_data(size_t skip_rows, size_t num_rows)
         out_buf.user_data |= PARQUET_COLUMN_BUFFER_FLAG_LIST_TERMINATED;
       } else if (out_buf.type.id() == type_id::STRING) {
         // need to cap off the string offsets column
-        size_type const sz = static_cast<size_type>(col_sizes[idx]);
+        size_type const sz = static_cast<size_type>(col_string_sizes[idx]);
         cudaMemcpyAsync(static_cast<int32_t*>(out_buf.data()) + out_buf.size,
```
Comment (bdice) [inline cpp/src/io/parquet/reader_impl.cpp:286]: This seems like a potential bug -- using `int32_t*` here and `sizeof(size_type)` below. Shouldn't these types match?
```suggestion
        cudaMemcpyAsync(static_cast<size_type*>(out_buf.data()) + out_buf.size,
```

**Review by bdice — APPROVED** [review]

File: cpp/src/io/parquet/reader_impl.cpp [hunk 3]
```diff
@@ -272,17 +273,17 @@ void reader::impl::decode_page_data(size_t skip_rows, size_t num_rows)
...
+                                      sizeof(size_type),
                                       cudaMemcpyDefault,
                                       _stream.value()));
         out_buf.user_data |= PARQUET_COLUMN_BUFFER_FLAG_LIST_TERMINATED;
       } else if (out_buf.type.id() == type_id::STRING) {
         // need to cap off the string offsets column
-        size_type const sz = static_cast<size_type>(col_sizes[idx]);
-        cudaMemcpyAsync(static_cast<int32_t*>(out_buf.data()) + out_buf.size,
+        auto const sz = static_cast<size_type>(col_string_sizes[idx]);
+        cudaMemcpyAsync(static_cast<size_type*>(out_buf.data()) + out_buf.size,
```
Comment (harrism) [inline cpp/src/io/parquet/reader_impl.cpp:286]: suggestion: I think this should be wrapped in `CUDF_CUDA_TRY`.

**Review by harrism — CHANGES_REQUESTED** [review]
Just one missed error handling macro. Otherwise looks good.

**Review by harrism — APPROVED** [review]
Thanks!

**Pull request merged**